Image Optimisation Pipeline

sbearcsiro opened the pull request:
- Add a configurable image optimisation pipeline
  - Calls a variety of external tools
  - Allows per image format sets of tools
  - Allows fallbacks to account for eg imagemagick v6 to v7, optipng -> oxipng
  - Allows Java based tools
  - Can convert formats of incoming images if wanted
- Extract embedded image metadata before optimisation in case stripped by tool
- Strip old comments from build.gradle

sbearcsiro committed:
Image Optimisation Pipeline
- Add a configurable image optimisation pipeline
  - Calls a variety of external tools
  - Allows per image format sets of tools
  - Allows fallbacks to account for eg imagemagick upgrades, optipng -> oxipng
  - Allows Java based tools
  - Can convert formats of incoming images if wanted
- Extract embedded image metadata before optimisation in case stripped by tool
- Strip old comments from build.gradle

vjrj commented on src/test/groovy/au/org/ala/images/ImageOptimisationServiceSpec.groovy:
```diff
@@ -0,0 +1,561 @@
+        given:
+        def service = new ImageOptimisationService()
+        service.commandExecutor = new FakeExec()
+        File tempDir = File.createTempDir('opt', 'spec')
```
Small issue, tempDir is not removed after the tests:

drwx------   3 vjrj   vjrj        4096 dic 18 17:25  optspec10247594337230432048
drwx------   3 vjrj   vjrj        4096 dic 18 17:25  optspec6684001193698184640
drwx------   3 vjrj   vjrj        4096 dic 18 17:25  optspec18151838450250103153
drwx------   3 vjrj   vjrj        4096 dic 18 17:25  optspec5944015659830252269
drwx------   4 vjrj   vjrj        4096 dic 18 17:25  optspec14803137498563332877
drwx------   3 vjrj   vjrj        4096 dic 18 17:25  optspec14477330968046229010
drwx------   3 vjrj   vjrj        4096 dic 18 17:25  optspec7543257582419710975
drwx------   2 vjrj   vjrj        4096 dic 18 17:25  optspec9078898221792595482
drwx------   3 vjrj   vjrj        4096 dic 18 17:25  optspec395602275564617550
drwx------   3 vjrj   vjrj        4096 dic 18 17:25  optspec5892751274077655397
drwx------   3 vjrj   vjrj        4096 dic 18 17:26  optspec6070546716765106793
drwx------   3 vjrj   vjrj        4096 dic 18 17:26  optspec17774677612034375626
drwx------   3 vjrj   vjrj        4096 dic 18 17:26  optspec17966754811671109431
drwx------   3 vjrj   vjrj        4096 dic 18 17:26  optspec8804592954123451379
drwx------   4 vjrj   vjrj        4096 dic 18 17:26  optspec3464889359782766923
drwx------   3 vjrj   vjrj        4096 dic 18 17:26  optspec5674045860961141501
drwx------   3 vjrj   vjrj        4096 dic 18 17:26  optspec1179015083508288133
drwx------   2 vjrj   vjrj        4096 dic 18 17:26  optspec4449851571508471063
drwx------   3 vjrj   vjrj        4096 dic 18 17:26  optspec4135259968006133507
drwx------   3 vjrj   vjrj        4096 dic 18 17:26  optspec3971438757400145217

vjrj commented on grails-app/conf/application.yml:
```diff
@@ -421,6 +421,260 @@ images:
+              - "0"
+              - "%IN%"
+              - "%OUT%"
+    tools:
```
It will be great to have a task that installs all these tools via apt in ala-instal (I can help here) to test this PR with real images.

sbearcsiro replied:
I've got some in here AtlasOfLivingAustralia/ala-install#930

vjrj replied:
Great. I'll test.

sbearcsiro committed:
Use @TempDir for ImageOptimisationServiceSpec

sbearcsiro committed:
Fixup extractedMetadata name in results map

sbearcsiro committed:
- Work around broken Grails + Spring Boot @Configuration with complex types
- Update call sites for withImageReader due to changes in image-utils
- Disable optimisation by default
- Add test for config loading

sbearcsiro committed:
- Update resize tool to use magick so that it falls back to convert if not found
- Add some debug logging to ImageOptimisationService
- Update test setup

sbearcsiro committed:
- Change ImageOptimisationConfigLoaderSpec to use Grails unit test doWithConfig instead of Mocks
- Remove @ConfigurationProperties annotation from ImageOptimisationConfig
- Update object property loading in ImageOptimisationConfigLoader

sbearcsiro committed:
- Remove @EnableConfigurationProperties(ImageOptimisationConfig) and create ImageOptimisationConfig as a bean

sbearcsiro committed:
- Fix wrong variable name in ImageService error handling
- Add info.app.name and version overrides for testing as the build system doesn't set them for integration tests
